### PR TITLE
Adding MSELoss, KLDivLoss and BCELoss to C++ front-end

### PR DESCRIPTION
**Pull request by ShahriarRezghi**
This PR adds ```MSELoss```, ```KLDivLoss``` and ```BCELoss```. The tests for ```BCELoss``` fail with the following error:
```
unknown file: Failure
C++ exception with description "autograd_meta() INTERNAL ASSERT FAILED at /home/shahriar/Contrib/pytorch/c10/core/TensorImpl.h:533, please report a bug to PyTorch. set_requires_grad is not implemented for Tensor (set_requires_grad at /home/shahriar/Contrib/pytorch/c10/core/TensorImpl.h:533)
```

### CODE REVIEW

File: torch/csrc/api/include/torch/nn/options/loss.h [hunk 1]
```diff
@@ -18,6 +18,21 @@ struct TORCH_API L1LossOptions {
 
 // ============================================================================
 
+/// Options for a KLDiv loss module.
+using KLDivLossOptions = L1LossOptions;
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/loss.h]: I think we should define all the new loss options explicitly instead of aliasing `L1LossOptions`, because `L1LossOptions` could take a different default `reduction` value or get a new options arg one day, and we might not want those changes to affect the other loss options.

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 2]
```diff
@@ -6,15 +6,33 @@ namespace torch {
...
+    const KLDivLossOptions& options) {
+  return torch::kl_div(self, target, options.reduction());
+}
+
+inline Tensor mse_loss(
+    const Tensor& self,
+    const Tensor& target,
+    const MSELossOptions& options) {
+  return torch::mse_loss(self, target, options.reduction());
+}
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/loss.h]: It would be awesome to add tests for the new functionals as well :D

File: torch/csrc/api/include/torch/nn/options/loss.h [hunk 3]
```diff
@@ -18,6 +18,21 @@ struct TORCH_API L1LossOptions {
...
+
+// ============================================================================
+
+/// Options for a MSE loss module.
+using MSELossOptions = L1LossOptions;
+
+// ============================================================================
+
+/// Options for a BCE loss module.
+using BCELossOptions = L1LossOptions;
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/loss.h]: I think `BCELossOptions` can take `weight` and `reduction` as constructor args, so it's not strictly equivalent to `L1LossOptions`.

File: torch/csrc/api/src/nn/modules/loss.cpp [hunk 4]
```diff
@@ -18,6 +17,46 @@ Tensor L1LossImpl::forward(const Tensor& input, const Tensor& target) {
...
 
+KLDivLossImpl::KLDivLossImpl(const KLDivLossOptions& options_)
+    : options(options_) {}
+
+void KLDivLossImpl::pretty_print(std::ostream& stream) const {
+  stream << "torch::nn::KLDivLoss";
+}
+
+Tensor KLDivLossImpl::forward(const Tensor& input, const Tensor& target) {
+  return torch::kl_div(input, target, options.reduction());
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/loss.cpp]: I think we should call `F::kl_div` here

File: torch/csrc/api/src/nn/modules/loss.cpp [hunk 5]
```diff
@@ -18,6 +17,46 @@ Tensor L1LossImpl::forward(const Tensor& input, const Tensor& target) {
...
+// ============================================================================
+
+MSELossImpl::MSELossImpl(const MSELossOptions& options_) : options(options_) {}
+
+void MSELossImpl::pretty_print(std::ostream& stream) const {
+  stream << "torch::nn::MSELoss";
+}
+
+Tensor MSELossImpl::forward(const Tensor& input, const Tensor& target) {
+  return torch::mse_loss(input, target, options.reduction());
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/loss.cpp]: ditto for `F::mse_loss`

File: torch/csrc/api/src/nn/modules/loss.cpp [hunk 6]
```diff
@@ -18,6 +17,46 @@ Tensor L1LossImpl::forward(const Tensor& input, const Tensor& target) {
...
+  register_parameter("weight", weight);
+}
+
+void BCELossImpl::pretty_print(std::ostream& stream) const {
+  stream << "torch::nn::BCELoss";
+}
+
+Tensor BCELossImpl::forward(const Tensor& input, const Tensor& target) {
+  return torch::binary_cross_entropy(
+      input, target, weight, options.reduction());
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/loss.cpp]: It would be fantastic to add `F::binary_cross_entropy` in this PR, and here we can call `F:: binary_cross_entropy`, to match the Python version better. :)

File: torch/csrc/api/src/nn/modules/loss.cpp [hunk 7]
```diff
@@ -18,6 +17,46 @@ Tensor L1LossImpl::forward(const Tensor& input, const Tensor& target) {
...
+}
+
+Tensor MSELossImpl::forward(const Tensor& input, const Tensor& target) {
+  return torch::mse_loss(input, target, options.reduction());
+}
+
+// ============================================================================
+
+BCELossImpl::BCELossImpl(const BCELossOptions& options_) : options(options_) {
+  register_parameter("weight", weight);
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/loss.cpp]:
```suggestion
  register_parameter("weight", options.weight());
```
which will also fix the `autograd_meta() INTERNAL ASSERT FAILED` error. :D

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 8]
```diff
@@ -6,15 +6,33 @@ namespace torch {
 namespace nn {
 namespace functional {
 
+inline Tensor l1_loss(
+    const Tensor& self,
+    const Tensor& target,
+    const L1LossOptions& options) {
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/loss.h]: I think we should do
```suggestion
    const L1LossOptions& options = {}) {
```
so that users can call this without passing options. And we should add a test for the `l1_loss(input, target)` use case too. (And same for all other functionals in this PR.)

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 9]
```diff
@@ -6,15 +6,33 @@ namespace torch {
...
+    const Tensor& self,
+    const Tensor& target,
+    const KLDivLossOptions& options) {
+  return torch::kl_div(self, target, options.reduction());
+}
+
+inline Tensor mse_loss(
+    const Tensor& self,
+    const Tensor& target,
+    const MSELossOptions& options) {
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/loss.h]: ditto for `options = {}`

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 10]
```diff
@@ -6,15 +6,33 @@ namespace torch {
...
+    const Tensor& self,
+    const Tensor& target,
+    const L1LossOptions& options) {
+  return torch::l1_loss(self, target, options.reduction());
+}
+
+inline Tensor kl_div(
+    const Tensor& self,
+    const Tensor& target,
+    const KLDivLossOptions& options) {
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/loss.h]: ditto for `options = {}`

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 11]
```diff
@@ -6,15 +6,33 @@ namespace torch {
 namespace nn {
 namespace functional {
 
+inline Tensor l1_loss(
+    const Tensor& self,
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/loss.h]: nit: `input` instead of `self`, to match Python version better :)

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 12]
```diff
@@ -6,15 +6,33 @@ namespace torch {
...
+    const Tensor& target,
+    const L1LossOptions& options) {
+  return torch::l1_loss(self, target, options.reduction());
+}
+
+inline Tensor kl_div(
+    const Tensor& self,
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/loss.h]: ditto: `self` -> `input`

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 13]
```diff
@@ -6,15 +6,33 @@ namespace torch {
...
+    const Tensor& target,
+    const KLDivLossOptions& options) {
+  return torch::kl_div(self, target, options.reduction());
+}
+
+inline Tensor mse_loss(
+    const Tensor& self,
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/loss.h]: ditto: `self` -> `input`

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 14]
```diff
@@ -6,15 +6,33 @@ namespace torch {
...
+    const Tensor& target,
+    const MSELossOptions& options) {
+  return torch::mse_loss(self, target, options.reduction());
+}
+
 inline Tensor hinge_embedding_loss(
-    const Tensor& x1,
-    const Tensor& x2,
+    const Tensor& self,
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/loss.h]: ditto: `self` -> `input`

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 15]
```diff
@@ -6,15 +6,33 @@ namespace torch {
...
+    const Tensor& target,
     const HingeEmbeddingLossOptions& options) {
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/loss.h]: ditto for `options = {}`

**Review by yf225 — COMMENTED** [review]
@ShahriarSS Thanks a lot for the great work and my sincere apologies for the delay. I left some comments.

File: torch/csrc/api/src/nn/modules/loss.cpp (hunk 7 above)
Comment (ShahriarSS) [reply torch/csrc/api/src/nn/modules/loss.cpp]: @yf225 it didn't fix the error.

Comment (yf225) [reply torch/csrc/api/src/nn/modules/loss.cpp]: @ShahriarSS Thanks for the catch! I think https://github.com/pytorch/pytorch/pull/27948 should fix it, and we can try to CI again after the fix.

File: torch/csrc/api/include/torch/nn/modules/loss.h [hunk 16]
```diff
@@ -37,6 +37,82 @@ struct TORCH_API L1LossImpl : public Cloneable<L1LossImpl> {
...
+/// See the documentation for `MSELossImpl` class to learn what methods it
+/// provides, or the documentation for `ModuleHolder` to learn about PyTorch's
+/// module storage semantics.
+TORCH_MODULE(MSELoss);
+
+// ============================================================================
+
+/// Creates a criterion that measures the Binary Cross Entropy
+/// between the target and the output.
+struct TORCH_API BCELossImpl : Module {
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/loss.h]: I think all `Impl` classes need to subclass from `public Cloneable<ImplName>` and implement the `void reset() override` method, otherwise `module->clone()` won't work on them.

File: torch/csrc/api/include/torch/nn/options/loss.h [hunk 17]
```diff
@@ -18,6 +18,44 @@ struct TORCH_API L1LossOptions {
...
+};
+
+// ============================================================================
+
+/// Options for a BCE loss module.
+struct TORCH_API BCELossOptions {
+  BCELossOptions(
+      Tensor weight = {},
+      Reduction::Reduction reduction = Reduction::Mean)
+      : weight_(weight), reduction_(reduction) {}
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/loss.h]: I think as a convention we should only provide a non-default constructor when the options has non-optional arguments or when the options has only one argument. For `BCELossOptions` I think we can follow the design of `HingeEmbeddingLossOptions` by providing defaults to `weight` and `reduction` and removing the non-default constructor.

File: torch/csrc/api/src/nn/modules/loss.cpp [hunk 18]
```diff
@@ -15,7 +14,46 @@ void L1LossImpl::pretty_print(std::ostream& stream) const {
...
+}
+
+// ============================================================================
+
+BCELossImpl::BCELossImpl(const BCELossOptions& options_) : options(options_) {
+  register_parameter("weight", options.weight());
+}
+
+void BCELossImpl::pretty_print(std::ostream& stream) const {
+  stream << "torch::nn::BCELoss";
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/loss.cpp]: It would be fantastic to add tests for the `pretty_print`s as well :D

File: torch/csrc/api/src/nn/modules/loss.cpp [hunk 19]
```diff
@@ -15,7 +14,46 @@ void L1LossImpl::pretty_print(std::ostream& stream) const {
...
+  return F::l1_loss(input, target, options);
+}
+
+// ============================================================================
+
+KLDivLossImpl::KLDivLossImpl(const KLDivLossOptions& options_)
+    : options(options_) {}
+
+void KLDivLossImpl::pretty_print(std::ostream& stream) const {
+  stream << "torch::nn::KLDivLoss";
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/loss.cpp]: To better match the Python version, we might need `torch::nn::KLDivLoss()`

File: torch/csrc/api/src/nn/modules/loss.cpp [hunk 20]
```diff
@@ -15,7 +14,46 @@ void L1LossImpl::pretty_print(std::ostream& stream) const {
...
+Tensor KLDivLossImpl::forward(const Tensor& input, const Tensor& target) {
+  return F::kl_div(input, target, options);
+}
+
+// ============================================================================
+
+MSELossImpl::MSELossImpl(const MSELossOptions& options_) : options(options_) {}
+
+void MSELossImpl::pretty_print(std::ostream& stream) const {
+  stream << "torch::nn::MSELoss";
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/loss.cpp]: To better match the Python version, we might need `torch::nn::MSELoss()`

File: torch/csrc/api/src/nn/modules/loss.cpp (hunk 18 above)
Comment (yf225) [reply torch/csrc/api/src/nn/modules/loss.cpp]: To better match the Python version, we might need `torch::nn::BCELoss()`

**Review by yf225 — COMMENTED** [review]
@ShahriarSS Thanks a lot for the awesome work! I left some minor comments.

**Comment by ShahriarSS** [comment]
@yf225 Should we do anything in ```BCELoss```'s reset function to ```options.weight()```?

File: torch/csrc/api/src/nn/modules/loss.cpp [hunk 21]
```diff
@@ -5,17 +5,61 @@ namespace F = torch::nn::functional;
...
+}
+
+Tensor MSELossImpl::forward(const Tensor& input, const Tensor& target) {
+  return F::mse_loss(input, target, options);
+}
+
+// ============================================================================
+
+BCELossImpl::BCELossImpl(const BCELossOptions& options_) : options(options_) {
+  register_parameter("weight", options.weight());
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/loss.cpp]: Thanks for the catch! Yes I think we should move this into `reset()`

Comment (yf225) [reply torch/csrc/api/src/nn/modules/loss.cpp]: Also it seems that it should be a buffer, not a parameter, based on the Python version:
```python
self.register_buffer('weight', weight)
```

**Review by yf225 — APPROVED** [review]
Thanks a lot for the awesome work @ShahriarSS!
